File: app/src/main/java/com/mygolbs/mybus/MainActivity.java
package com.mygolbs.mybus;

import android.app.Activity;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.ScrollView;
import android.widget.TextView;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.URLEncoder;

public class MainActivity extends Activity {
    private static final String LINE_API = "http://quanguo.mygolbs.com:8081/Lostfound/GetLineServlet";
    private EditText cityCodeInput;
    private EditText lineInput;
    private TextView resultView;
    private ProgressBar progressBar;
    private Button searchButton;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        buildUi();
    }

    private void buildUi() {
        LinearLayout root = new LinearLayout(this);
        root.setOrientation(LinearLayout.VERTICAL);
        int pad = dp(14);
        root.setPadding(pad, pad, pad, pad);

        TextView title = new TextView(this);
        title.setText("掌上公交真实 API 调试版");
        title.setTextSize(20);
        root.addView(title, new LinearLayout.LayoutParams(-1, -2));

        cityCodeInput = new EditText(this);
        cityCodeInput.setHint("城市代码：武汉=02700");
        cityCodeInput.setSingleLine(true);
        cityCodeInput.setText("02700");
        root.addView(cityCodeInput, new LinearLayout.LayoutParams(-1, -2));

        lineInput = new EditText(this);
        lineInput.setHint("线路，如 5路");
        lineInput.setSingleLine(true);
        lineInput.setText("5路");
        root.addView(lineInput, new LinearLayout.LayoutParams(-1, -2));

        searchButton = new Button(this);
        searchButton.setText("查询武汉线路 JSON");
        root.addView(searchButton, new LinearLayout.LayoutParams(-1, -2));

        progressBar = new ProgressBar(this);
        progressBar.setVisibility(View.GONE);
        root.addView(progressBar, new LinearLayout.LayoutParams(-1, -2));

        ScrollView scroll = new ScrollView(this);
        resultView = new TextView(this);
        resultView.setTextSize(12);
        resultView.setText("已验证真实 JSON 接口：\nPOST " + LINE_API + "\n参数：cityCode=02700\n\n点击查询会过滤 5路。\n注意：MyBusWeChatMessage 返回 HTML，不是公交 JSON。");
        scroll.addView(resultView, new ScrollView.LayoutParams(-1, -2));
        root.addView(scroll, new LinearLayout.LayoutParams(-1, 0, 1));

        searchButton.setOnClickListener(v -> search());
        setContentView(root);
    }

    private void search() {
        final String cityCode = cityCodeInput.getText().toString().trim();
        final String keyword = lineInput.getText().toString().trim();
        progressBar.setVisibility(View.VISIBLE);
        searchButton.setEnabled(false);
        resultView.setText("请求中...");
        new Thread(() -> {
            String out;
            try {
                String body = requestLines(cityCode);
                out = filterLines(body, keyword);
            } catch (Exception e) {
                out = "请求失败: " + e.getMessage();
            }
            final String finalOut = out;
            runOnUiThread(() -> {
                progressBar.setVisibility(View.GONE);
                searchButton.setEnabled(true);
                resultView.setText(finalOut);
            });
        }).start();
    }

    private String requestLines(String cityCode) throws Exception {
        String post = "cityCode=" + URLEncoder.encode(cityCode, "UTF-8");
        HttpURLConnection c = (HttpURLConnection) new URL(LINE_API).openConnection();
        c.setConnectTimeout(10000);
        c.setReadTimeout(15000);
        c.setRequestMethod("POST");
        c.setDoOutput(true);
        c.setRequestProperty("User-Agent", "Mozilla/5.0 MyBus-Lite");
        c.setRequestProperty("Referer", "http://quanguo.mygolbs.com:8081/Lostfound/");
        c.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
        OutputStream os = c.getOutputStream();
        os.write(post.getBytes("UTF-8"));
        os.close();
        int code = c.getResponseCode();
        BufferedReader r = new BufferedReader(new InputStreamReader(
                code >= 200 && code < 400 ? c.getInputStream() : c.getErrorStream(), "UTF-8"));
        StringBuilder sb = new StringBuilder();
        String line;
        while ((line = r.readLine()) != null) sb.append(line).append('\n');
        r.close();
        c.disconnect();
        return "URL: " + LINE_API + "\nPOST: " + post + "\nHTTP: " + code + "\n\n" + sb;
    }

    private String filterLines(String raw, String keyword) {
        if (keyword == null || keyword.length() == 0) return pretty(raw);
        StringBuilder out = new StringBuilder();
        out.append("关键词：").append(keyword).append("\n\n");
        String[] parts = raw.replace("},{", "}\n{").split("\n");
        for (String part : parts) {
            if (part.contains(keyword)) out.append(part).append("\n\n");
        }
        if (out.toString().trim().equals("关键词：" + keyword)) {
            out.append("未过滤到，下面显示原始返回：\n").append(pretty(raw));
        }
        return out.toString();
    }

    private String pretty(String s) {
        return s.replace("},{", "},\n{").replace(",\"", ",\n\"");
    }

    private int dp(int v) { return (int) (v * getResources().getDisplayMetrics().density + 0.5f); }
}

File: app/src/main/java/com/mygolbs/mybus/MainActivity.kt
package com.mygolbs.mybus

import android.os.Bundle
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder

/**
 * 掌上公交查询 - 使用真实掌上公交 API
 *
 * 从 www.mygolbs.com 的 app.js 中提取的 API：
 * - WxBusServer: https://wx.mygolbs.com/WxBusServer/
 * - MyBusWeChatMessage: https://wx.mygolbs.com/MyBusWeChatMessage
 */
class MainActivity : AppCompatActivity() {

    private lateinit var etLine: EditText
    private lateinit var etCity: EditText
    private lateinit var btnSearch: Button
    private lateinit var tvResult: TextView
    private lateinit var pbLoading: ProgressBar

    companion object {
        // 核心 API
        private const val BASE_URL = "https://wx.mygolbs.com"
        private const val API_PATH = "/WxBusServer/"
        private const val DATA_PATH = "/MyBusWeChatMessage"
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        etLine = findViewById(R.id.et_line)
        etCity = findViewById(R.id.et_city)
        btnSearch = findViewById(R.id.btn_search)
        tvResult = findViewById(R.id.tv_result)
        pbLoading = findViewById(R.id.pb_loading)

        // 默认城市：福州
        etCity.setText("福州")

        btnSearch.setOnClickListener {
            val line = etLine.text.toString().trim()
            val city = etCity.text.toString().trim()
            if (line.isEmpty()) {
                tvResult.text = "请输入公交线路名"
                return@setOnClickListener
            }
            searchBus(line, city)
        }
    }

    private fun searchBus(line: String, city: String) {
        pbLoading.visibility = ProgressBar.VISIBLE
        tvResult.text = "正在查询…"
        btnSearch.isEnabled = false

        Thread {
            try {
                val result = queryApi(line, city)
                runOnUiThread {
                    tvResult.text = result
                    pbLoading.visibility = ProgressBar.GONE
                    btnSearch.isEnabled = true
                }
            } catch (e: Exception) {
                runOnUiThread {
                    tvResult.text = "错误: ${e.message}\n\n${queryFallback(line, city)}"
                    pbLoading.visibility = ProgressBar.GONE
                    btnSearch.isEnabled = true
                }
            }
        }.start()
    }

    /** 调用掌上公交 API（WxBusServer） */
    private fun queryApi(line: String, city: String): String {
        // 先从 WeChatMessage 接口查询
        val url = "$BASE_URL$DATA_PATH?cmd=204&" +
                "city=${URLEncoder.encode(city, "UTF-8")}&" +
                "line=${URLEncoder.encode(line, "UTF-8")}&" +
                "output=json"

        val conn = URL(url).openConnection() as HttpURLConnection
        conn.connectTimeout = 5000
        conn.readTimeout = 5000
        conn.setRequestProperty("Content-Type", "application/json;charset=utf-8")
        conn.requestMethod = "GET"

        val code = conn.responseCode
        if (code != 200) {
            return "API 返回 $code\n\n尝试备用接口…\n${queryWs(line, city)}"
        }

        val reader = BufferedReader(InputStreamReader(conn.inputStream, "UTF-8"))
        val sb = StringBuilder()
        var l: String?
        while (reader.readLine().also { l = it } != null) sb.append(l)
        reader.close()
        conn.disconnect()

        if (sb.isEmpty()) return "无数据"
        return formatJson(sb.toString())
    }

    /** 备用接口（WxBusServer） */
    private fun queryWs(line: String, city: String): String {
        val url = "$BASE_URL$API_PATH?cmd=205&" +
                "city=${URLEncoder.encode(city, "UTF-8")}&" +
                "line=${URLEncoder.encode(line, "UTF-8")}&" +
                "output=json"

        try {
            val conn = URL(url).openConnection() as HttpURLConnection
            conn.connectTimeout = 3000
            conn.readTimeout = 3000
            conn.requestMethod = "GET"

            val reader = BufferedReader(InputStreamReader(conn.inputStream, "UTF-8"))
            val sb = StringBuilder()
            var l: String?
            while (reader.readLine().also { l = it } != null) sb.append(l)
            reader.close()
            conn.disconnect()

            return "备用接口:\n${formatJson(sb.toString())}"
        } catch (e: Exception) {
            return "备用接口失败: ${e.message}"
        }
    }

    private fun formatJson(str: String): String {
        // 简单格式化
        return try {
            str.replace("},{", "},\n{")
                .replace("[", "[\n").replace("]", "\n]")
                .replace(",", ",\n")
        } catch (e: Exception) { str }
    }
}